src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -422,7 +441,36 @@ private static RecursivePatternSyntax RecursivePattern(SubpatternSyntax subpatte
                 commonReceiver = GetInnermostReceiver(left, lastName, static (identifierName, lastName) => identifierName != lastName);
             }
 
-            // If the common receiver is null, it's an implicit `this` reference in source.
+            // If the common receiver is null, there might still be one in cases like these:
+            // `MyClassField != null && MyClassField.prop != 0`. In this case, the left expression doesn't say
+            // anything new to the second one so it should be discarded, but MyClassField should still act as the
+            // receiver instead of the implicit this so we get
+            // `MyClassField is { prop: not 0 }` instead of `this is { MyClassField: not null, MyClassField.prop: not 0 }`
+            // We need to cover this case for either side of the expression by detecting a null check on either side
+            if (AreEquivalent(leftNames[^1], rightNames[^1]))
```

Comment:
how do we know these are both non-empty? is it a guarantee of TryGetInnermostReceiver?

Reply:
Not the method itself, but yes in this context because both calls above this line needs to return true in order to not get an early return and for that to happen, they need to not be empty.

So we do have this guarantee here.